plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerPageSource.java

```diff
@@ -0,0 +1,707 @@
+
+class FakerPageSource
+        implements ConnectorPageSource
+{
+    static final long[] POWERS_OF_TEN = {
+            1L,
+            10L,
+            100L,
+            1_000L,
+            10_000L,
+            100_000L,
+            1_000_000L,
+            10_000_000L,
+            100_000_000L,
+            1_000_000_000L,
+            10_000_000_000L,
+            100_000_000_000L,
+            1_000_000_000_000L,
+            10_000_000_000_000L,
+            100_000_000_000_000L,
+            1_000_000_000_000_000L,
+            10_000_000_000_000_000L,
+            100_000_000_000_000_000L,
+            1_000_000_000_000_000_000L
+    };
+
+    private static final int ROWS_PER_REQUEST = 4096;
+
+    private final Random random;
+    private final Faker faker;
+    private final FakerTableHandle table;
+    private final List<BiConsumer<BlockBuilder, Long>> generators;
+    private long completedBytes;
+    private long completedRows;
+
+    private final PageBuilder pageBuilder;
+
+    private boolean closed;
+
+    FakerPageSource(Faker faker, Random random, List<FakerColumnHandle> columns, FakerTableHandle table)
+    {
+        this.faker = requireNonNull(faker, "faker is null");
+        this.random = requireNonNull(random, "random is null");
+        this.table = requireNonNull(table, "table is null");
+        List<Type> types = requireNonNull(columns, "columns is null")
+                .stream()
+                .map(FakerColumnHandle::type)
+                .collect(toList());
+        this.generators = columns
+                .stream()
+                .map(column -> constraintedValueGenerator(
+                        column,
+                        table.constraint().getDomains().get().getOrDefault(column, Domain.all(column.type()))))
+                .collect(toList());
+        // TODO create generators for every column
+        this.pageBuilder = new PageBuilder(types);
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return completedBytes;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return closed && pageBuilder.isEmpty();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        if (!closed) {
+            int positions = (int) Math.min(table.limit() - completedRows, ROWS_PER_REQUEST);
+            if (positions <= 0) {
+                closed = true;
+            }
+            else {
+                pageBuilder.declarePositions(positions);
+                for (int column = 0; column < generators.size(); column++) {
+                    BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(column);
+                    BiConsumer<BlockBuilder, Long> generator = generators.get(column);
+                    for (int i = 0; i < positions; i++) {
+                        generator.accept(blockBuilder, completedRows + i);
+                    }
```

Comment:
Ideally what we want is
```
generator.accept(blockBuilder, completedRows, positions);
```
Generate N positions in a batch.
This is a megamorphic call site, so looping over it per position can be slow

Reply:
I don't know what's a megamorphic call site, can you share some references? This would change the interfaces a lot, and I'm not confident I can do this correctly, without understanding the optimization. Is this a blocker for this PR?

Reply:
BTW I was benchmarking it like this, comparing to the `sequence` table function, and it was fast enough. I'm not sure what gains we can expect here.

Reply:
The short version is that when you have a method call that can be dispatched to more than 2 possible implementations, the JIT may not do certain optimizations and the code may end up being much slower than what it would have been had there been just 1 or 2 possible implementations.
Longer version in http://www.insightfullogic.com/2014/May/12/fast-and-megamorphic-what-influences-method-invoca/
To observe the per impact, you have to execute this code for 3 different data types and compare the throughput with what you would get if you only executed it on one data type.
This can be tackled in a follow-up, not important enough for current PR to tackle.